File: optional/typescript-workflow-runner/src/application/jobs/database/item-crud.ts
#!/usr/bin/env node
/**
 * Item CRUD Sample Job
 *
 * Demonstrates basic CRUD operations on SQLite items table.
 * Creates, reads, updates, and deletes sample items.
 *
 * Usage:
 *   npm run item-crud
 *   tsx src/jobs/database/item-crud.ts
 *
 * @module jobs/database/item-crud
 */

import "../../../infra/config/yaml.js";
import { createJobWithDeps } from "../../job-factory.js";
import { registerJob } from "../../../registry/job-registry.js";
import { runIfMain } from "../../../runner/workflow-runner.js";

/**
 * Item CRUD sample job
 *
 * Demonstrates all CRUD operations:
 * - CREATE: Insert a new item
 * - READ: Retrieve an item by ID
 * - UPDATE: Modify an existing item
 * - LIST: Get all items
 * - DELETE: Remove an item
 */
export const itemCrudJob = createJobWithDeps(
  {
    name: "item-crud",
    description: "Demonstrates CRUD operations on items table",
    schedule: "0 0 * * *", // Daily at midnight
  },
  async (_ctx, logger, deps) => {
    const { itemRepo, clock } = deps;
    const operations: string[] = [];

    // CREATE
    logger.info("Creating sample item...");
    const created = itemRepo.create({
      name: `Sample Item ${clock.toISOString()}`,
    });
    operations.push(`CREATE: id=${created.id}, name=${created.name}`);
    logger.info({ item: created }, "Created item");

    // READ
    logger.info({ id: created.id }, "Reading item...");
    const retrieved = itemRepo.get(created.id);
    if (retrieved) {
      operations.push(`READ: id=${retrieved.id}, name=${retrieved.name}`);
      logger.info({ item: retrieved }, "Retrieved item");
    }

    // UPDATE
    logger.info({ id: created.id }, "Updating item...");
    const updated = itemRepo.update(created.id, {
      name: `Updated Item ${clock.toISOString()}`,
    });
    if (updated) {
      operations.push(`UPDATE: id=${updated.id}, name=${updated.name}`);
      logger.info({ item: updated }, "Updated item");
    }

    // LIST
    logger.info("Listing all items...");
    const allItems = itemRepo.list();
    operations.push(`LIST: ${allItems.length} items total`);
    logger.info({ count: allItems.length }, "Listed items");

    // DELETE
    logger.info({ id: created.id }, "Deleting item...");
    const deleted = itemRepo.delete(created.id);
    operations.push(`DELETE: id=${created.id}, success=${deleted}`);
    logger.info({ deleted }, "Deleted item");

    // Final count
    const finalCount = itemRepo.list().length;

    return {
      success: true,
      message: `CRUD operations completed. ${finalCount} items remaining.`,
      metrics: {
        operationsRun: operations.length,
        itemsRemaining: finalCount,
      },
      details: { operations },
    };
  }
);

registerJob(itemCrudJob);
export const main = itemCrudJob.handler;
runIfMain(main, import.meta.url);

File: starters/rust-batch/src/application/jobs/item_crud.rs
//! Item CRUD Sample Job
//!
//! Demonstrates basic CRUD operations on SQLite items table.
//! Creates, reads, updates, and deletes sample items.
//!
//! Usage:
//!   cargo run -- --job item-crud

use std::sync::Arc;

use serde_json::json;
use tracing::info;

use crate::application::job_factory::{create_job, JobDefinition, JobOptions};
use crate::application::job_types::JobResult;
use crate::di::AppContext;
use crate::domain::{NewItem, UpdateItem};

/// Create the item-crud job
pub fn item_crud_job() -> JobDefinition {
    create_job(
        JobOptions::new("item-crud", "Demonstrates CRUD operations on items table")
            .with_schedule("0 0 * * *"), // Daily at midnight
        |_ctx, app_ctx: Arc<AppContext>| async move {
            let mut operations: Vec<String> = Vec::new();

            // CREATE
            info!("Creating sample item...");
            let created = app_ctx.item_repo.create(NewItem {
                name: format!("Sample Item {}", app_ctx.clock.to_iso_string()),
            })?;
            operations.push(format!("CREATE: id={}, name={}", created.id, created.name));
            info!(item_id = %created.id, item_name = %created.name, "Created item");

            // READ
            info!(id = %created.id, "Reading item...");
            if let Some(retrieved) = app_ctx.item_repo.get(created.id)? {
                operations.push(format!(
                    "READ: id={}, name={}",
                    retrieved.id, retrieved.name
                ));
                info!(item_id = %retrieved.id, item_name = %retrieved.name, "Retrieved item");
            }

            // UPDATE
            info!(id = %created.id, "Updating item...");
            if let Some(updated) = app_ctx.item_repo.update(
                created.id,
                UpdateItem {
                    name: Some(format!("Updated Item {}", app_ctx.clock.to_iso_string())),
                },
            )? {
                operations.push(format!("UPDATE: id={}, name={}", updated.id, updated.name));
                info!(item_id = %updated.id, item_name = %updated.name, "Updated item");
            }

            // LIST
            info!("Listing all items...");
            let all_items = app_ctx.item_repo.list()?;
            operations.push(format!("LIST: {} items total", all_items.len()));
            info!(count = %all_items.len(), "Listed items");

            // // DELETE
            // info!(id = %created.id, "Deleting item...");
            // let deleted = app_ctx.item_repo.delete(created.id)?;
            // operations.push(format!("DELETE: id={}, success={}", created.id, deleted));
            // info!(deleted = %deleted, "Deleted item");

            // Final count
            let final_count = app_ctx.item_repo.list()?.len() as i64;

            Ok(JobResult::success(format!(
                "CRUD operations completed. {} items remaining.",
                final_count
            ))
            .with_metric("operations_run", operations.len() as i64)
            .with_metric("items_remaining", final_count)
            .with_details(json!({ "operations": operations })))
        },
    )
}
